sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala

```diff
@@ -498,7 +498,7 @@ class Dataset[T] private[sql](
    * @since 1.6.0
    */
   def explain(extended: Boolean): Unit = {
-    val explain = ExplainCommand(queryExecution.logical, extended = extended)
+    val explain = ExplainCommand(queryExecution, extended = extended)
```

Comment:
To fix the current bug, I agree with @viirya . I believe this is inevitable. 
Could you give us some guide for this, @cloud-fan and @gatorsmile ?